### PR TITLE
Add a page for data modeling with ScalarDB

### PR DESCRIPTION
**Pull request by feeblefakie**
## Description

This PR adds a page for ScalarDB data modeling because there are no pages explaining it.

## Related issues and/or PRs

N/A

## Changes made

- Added explanation about the ScalarDB data model.
- Added explanation about how to design database schemas

## Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes.
- [ ] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [ ] Tests (unit, integration, etc.) have been added for the changes.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

> Provide any additional information or notes that may be relevant to the reviewers or stakeholders.

## Release notes

Added a page for data modeling with ScalarDB.

### CODE REVIEW

File: docs/data-modeling.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+A table is a collection of data. A namespace most often contains one or more tables, each identified by a name. Each table includes a set of partitions.
+
+#### Partition
+A partition is a collection of records and a unit of (logical) distribution to nodes. Therefore, records within the same partition are placed in the same logical location.
+
+#### Record / Row
+A record or row is a set of columns that is uniquely identifiable among all of the other records.
+
+#### Column
+A column is a fundamental data element and does not need to be broken down any further. Each record is composed of one or more columns. Each column has a data type. For details about the data type, refer to [Data-type mapping between ScalarDB and other databases](https://scalardb.scalar-labs.com/docs/latest/schema-loader#data-type-mapping-between-scalardb-and-other-databases).
```
Comment (feeblefakie) [inline docs/data-modeling.md:29]: A question to @josh-wong.
For this line, there is a link to another doc, but how should I do this?

File: docs/data-modeling.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#### Clustering key
+A clustering key uniquely identifies a record within a partition. It comprises a set of columns called clustering key columns. When you want to specify a clustering key, you should specify a partition key for efficient lookups. When you specify a clustering key without a partition key,
+you end up scanning all the partitions if you don’t have proper indexes. Scanning all the partitions is very time-consuming, so please do it only when you know what you are doing.
+
+Records within a partition are assumed to be sorted by clustering key columns. Therefore, you can specify a part of clustering key columns in the defined order to narrow down the results to be returned. 
+
+#### Index key
+An index key identifies records by looking up the key in indexes.
+
+## How to design your database schemas
```
Comment (feeblefakie) [inline docs/data-modeling.md:53]: I feel these are not enough. Please suggest anything for here!

File: docs/data-modeling.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+A partition key uniquely identifies a partition. A partition key comprises a set of columns, which are called partition key columns. When you specify only a partition key, you can get a set of records that belong to the partition.
+
+#### Clustering key
+A clustering key uniquely identifies a record within a partition. It comprises a set of columns called clustering key columns. When you want to specify a clustering key, you should specify a partition key for efficient lookups. When you specify a clustering key without a partition key,
+you end up scanning all the partitions if you don’t have proper indexes. Scanning all the partitions is very time-consuming, so please do it only when you know what you are doing.
+
+Records within a partition are assumed to be sorted by clustering key columns. Therefore, you can specify a part of clustering key columns in the defined order to narrow down the results to be returned. 
+
+#### Index key
+An index key identifies records by looking up the key in indexes.
```
Comment (komamitsu) [inline docs/data-modeling.md:51]: The relationship between Primary key (partition and clustering keys) and Index key doesn't seem to be described enough to me. I think it may be worth mentioning here that index key can be used as an alternative to primary key.

File: docs/data-modeling.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+The following diagram shows an example of ScalarDB tables, each a collection of records. This section first explains what objects, such as tables and records, ScalarDB defines and then describes how to locate records.
+
+![ScalarDB data model.](images/scalardb_data_model.png)
+
+### Objects in ScalarDB
+The ScalarDB data model has several objects.
+
+#### Namespace
+A namespace is a collection of tables analogous to an SQL namespace or database. Typically, each application creates all its tables in one namespace.
```
Comment (komamitsu) [inline docs/data-modeling.md:17]: I think another characteristic of ScalarDB's Namespace is that an arbitrary underlying database/storage can be used for each Namespace by leveraging the multi storage feature. I'm not sure the topic is out of the scope of this document, though.

File: docs/data-modeling.md (hunk 2 above)
Comment (komamitsu) [reply docs/data-modeling.md:53]: This might be out of the scope of this document. But, if we can mention the multi storage feature, putting write heavy tables in a namespace that uses a NoSQL database might be an interesting design topic.

File: docs/data-modeling.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+A partition is a collection of records and a unit of (logical) distribution to nodes. Therefore, records within the same partition are placed in the same logical location.
+
+#### Record / Row
+A record or row is a set of columns that is uniquely identifiable among all of the other records.
+
+#### Column
+A column is a fundamental data element and does not need to be broken down any further. Each record is composed of one or more columns. Each column has a data type. For details about the data type, refer to [Data-type mapping between ScalarDB and other databases](https://scalardb.scalar-labs.com/docs/latest/schema-loader#data-type-mapping-between-scalardb-and-other-databases).
+
+#### Index
+An index is a copy of the records in a single table, sorted by a column. ScalarDB queries use indexes to find data more efficiently in a table, given the values of a particular column.
```
Comment (jnmt) [inline docs/data-modeling.md:32]: I feel something wrong with the expression "a copy of the records" since it's not a whole copy of the record. It might be something like "a copy of a certain single column that has (logical?) links for the original records".

File: docs/data-modeling.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+#### Index
+An index is a copy of the records in a single table, sorted by a column. ScalarDB queries use indexes to find data more efficiently in a table, given the values of a particular column. 
+
+
+### How to locate records
+This section discusses how to locate records from a table.
+
+#### Primary key
+A primary key uniquely identifies each record; no two records can have the same primary key. Therefore, you can locate a record by specifying a primary key. A primary key comprises a set of partition keys and clustering keys.
```
Comment (jnmt) [inline docs/data-modeling.md:39]: The following might be better since I feel a partition key is not a set of keys but a set of columns, according to the partition and clustering key description.

```suggestion
A primary key uniquely identifies each record; no two records can have the same primary key. Therefore, you can locate a record by specifying a primary key. A primary key comprises a partition key and, optionally, a clustering key.
```

File: docs/data-modeling.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+#### Index key
+An index key identifies records by looking up the key in indexes.
+
+## How to design your database schemas
+You can design your database schemas similarly to the relational model, but there are a few best practices to follow.
+
+### Care about data distribution
+It would be best if you tried to balance loads to partitions by properly selecting partition and clustering keys. For example, in a banking application, if you choose an account ID as a partition key, you can perform any account operations for a specific account within the partition to which the account belongs. So, if you operate on different account IDs, you access different partitions. On the other hand, if you choose a branch ID as a partition key and an account ID as a clustering key, all the accesses to a branch’s account IDs go to the same partition. In addition, you should choose a high cardinality column as a partition key because creating a small number of large partitions causes an imbalance in loads and data sizes.
```
Comment (jnmt) [inline docs/data-modeling.md:57]: > all the accesses to a branch’s account IDs go to the same partition.

Are you writing this part to say that it's not good (for most cases)?  If so, it would be nice if it could be more direct.

File: docs/data-modeling.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+### Try to read a single partition as much as possible
+Because of the data model characteristics, single partition lookup is most efficient. If you need to issue a scan or select request that requires multi-partition lookups or scans ([with cross-partition scan enabled](https://scalardb.scalar-labs.com/docs/latest/configurations/#cross-partition-scan-configurations)), make sure you know what you are doing and think about updating the schemas if possible. For example, in a banking application, if you choose email as a partition key and an account ID as a clustering key and issue a query that specifies an account ID, the query will span to all the partitions because it cannot identify the corresponding partition efficiently. In such a case, you should always look up the table with an account ID.
```
Comment (jnmt) [inline docs/data-modeling.md:60]: As we discussed a similar topic recently, should we avoid using even index access as much as possible if we can achieve the desired data access by exploiting only the partition and clustering key since the index access would cause cross-partition access?

File: docs/data-modeling.md (hunk 2 above)
Comment (jnmt) [inline docs/data-modeling.md:53]: This might be out of the scope, but we could add some tips to avoid the limitation of SclarDB specification. For instance, how should we create an index for multiple columns? Currently, we do not support it, but users can achieve a similar thing by creating an actual table that has links to the desired table.

**Review by jnmt — COMMENTED** [review]
This is a nice document for starters. Thank you. I left a few comments, so PTAL when you get a chance.

**Review by brfrn169 — COMMENTED** [review]
Left a comment for now.

I think we should discuss the "How to design your database schemas" part further. Let me take a closer look.

File: docs/data-modeling.md (hunk 4 above)
Comment (yu2scalar) [reply docs/data-modeling.md:17]: I agree with @komamitsu .
My impression of Namespace is that it helps abstract multiple databases as if they were a single database.

Comment (feeblefakie) [reply docs/data-modeling.md:17]: Thanks! Namespace mapping things should probably be described in the multi-storage transaction docs to make this doc more focused.
But, since ScalarDB tends to manage multiple namespaces, mapped from different databases, I will remove `Typically, each application creates all its tables in one namespace`.

File: docs/data-modeling.md (hunk 5 above)
Comment (feeblefakie) [reply docs/data-modeling.md:32]: Right, it's a description for clustered indexes.
ScalarDB indexes are unclustered indexes, so I will update the description.
Thanks!

**Comment by feeblefakie** [comment]
@brfrn169 @komamitsu @Torch3333 @jnmt @josh-wong 
I updated the PR based on your feedback.
It was kind of hard to update it for each feedback, so it would be great if you could take a look again.
I wouldn't like to spend too much time on the first version since this kind of doc is currently missing.
So, if it isn't wrong or too misleading, please accept it with some feedback. 
We can work on revising it in a later PR.

File: docs/data-modeling.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+As an alternative to secondary indexes, you can create another table that works as a (clustered) index of a base table. For example, assume there is a table with 3 columns, `table1(A, B, C)`, with a primary key `A`. Then, you can create a table like `index-table1(C, A, B)` with `C` as a primary key so that you can look up a single partition by specifying a value for `C`.
+This approach could speed up read queries but create more load to write queries because it writes to two tables with ScalarDB transactions.
+
+:::note
+
+We plan to have a table-based secondary index feature in the future.
+
+:::
+
+#### Care about data is (assumed to be) distributed by hashing
```
Comment (josh-wong) [inline docs/data-modeling.md:88]: The part in parentheses is confusing, so how about we reword this as follows?:

```suggestion
#### Consider how data is assumed to be distributed by hashing
```

File: docs/data-modeling.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+In the current ScalarDB data model, data is assumed to be distributed by hashing, so you can't perform range queries efficiently without a partition key.
+If you want to issue range queries efficiently, you need to do it within a partition.
+However, it's important to note that this approach requires you to specify a partition key. This can pose scalability issues as the range queries always go to the same partition, potentially overloading it.
+The limitation is not specific to ScalarDB but to databases where data is distributed by hashing for scalability.
```
Comment (josh-wong) [inline docs/data-modeling.md:93]:
```suggestion
In the current ScalarDB data model, data is assumed to be distributed by hashing. Therefore, you can't perform range queries efficiently without a partition key.

If you want to issue range queries efficiently, you need to do so within a partition. However, if you follow this approach, you must specify a partition key. This can pose scalability issues as the range queries always go to the same partition, potentially overloading it. This limitation is not specific to ScalarDB but to databases where data is distributed by hashing for scalability.
```

**Review by josh-wong — CHANGES_REQUESTED** [review]
I've added some comments and suggestions. PTAL!

**Review by komamitsu — APPROVED** [review]
LGTM! 👍

**Review by Torch3333 — APPROVED** [review]
LGTM, thank you!

**Review by jnmt — APPROVED** [review]
LGTM! Thank you!

File: docs/data-modeling.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+#### Index key
+An index key identifies records by looking up the key in indexes. An index key lookup spans all the partitions, so it is not necessarily very efficient, especially if the selectivity of a lookup is not low.
+
+## How to design your database schemas
+You can design your database schemas similarly to the relational model, but there is a basic principle and are a few best practices to follow.
+
+### Query-driven data modeling
+In relational databases, data is organized in normalized tables with foreign keys used to reference related data in other tables. The queries that the application will make are structured by the tables, and related data are queried as table joins.
+
+In ScalarDB, although it supports join operations, data modeling should be more query-driven, like NoSQL databases. The data access patterns and application queries should determine the structure and organization of tables.
```
Comment (brfrn169) [inline docs/data-modeling.md:60]: We supports join operations only in ScalarDB SQL. It might be better to mention that.

**Review by brfrn169 — APPROVED** [review]
Left a minor comment. But, LGTM! Thank you!
